### PR TITLE
add tracing support with custom roundtripper

### LINKED ISSUE #80: Add spn_ctx to header
Adding spn_ctx to event metadata is discussed in this [issue](https://github.com/stoewer/go-nakadi/issues/76). The span_context in the event is to correlate the event production with what the event consumers are doing. However, it is also possible to add spn_ctx to [headers](https://nakadi.io/manual.html#/event-types/name/events_post*span_ctx). The HTTP header is for when you need to correlate the process submitting the event with the processing on Nakadi side. It is useful to integrate Nakadi as a part of your trace.

### PR DESCRIPTION
**Pull request by rdilruba**
closes #80

### CODE REVIEW

**Comment by rdilruba** [comment]
Would it be possible to get a review for this pr? 🙋‍♀️ @stoewer

**Review by stoewer — CHANGES_REQUESTED** [review]
Thank you for reporting #80 and this PR. The PR looks already quite good, but I still think there are some things that should be improved.

Since the tracer is already implemented as `http.RoundTripper` middleware, it would be good to use this as an opportunity to extend the `ClientOptions` such that also other middleware can be used. This would further decouple the tracing implementation from the rest of the client and allows users to create their own tracing or logging solutions. 

The changed API might look like this:

1. Add middleware to the `ClientOptions` in `nakadi.go`:
```go
// Middleware provides a chainable http.RoundTripper middleware that can be used
// to hook into requests e.g. for logging or tracing purposes.
type Middleware func(transport http.RoundTripper) http.RoundTripper

// ClientOptions contains all non mandatory parameters used to instantiate the Nakadi client.
type ClientOptions struct {
	TokenProvider     func() (string, error)
	ConnectionTimeout time.Duration
	Middleware        Middleware
}
```
2. Make the `Transport` from `helper.go` a `Middleware` and move it to its own file, for example `tracing.go`:
```go
type TracingOptions struct {
	...
}

type TracingMiddleware struct {
	tr http.RoundTripper
	...
}

func (t TracingMiddleware)  RoundTrip(req *http.Request) (*http.Response, error) {
	...
}

func NewTracingMiddleware(options *TracingOptions) Middleware {
	return func(trandport http.RoundTripper) http.RoundTripper {
		return TracingMiddleware{
			tr: trandport,
			...
		}
	}
}
```
3. Use the middleware in `newHTTPClient`

File: helper.go [hunk 1]
```diff
@@ -18,38 +21,125 @@ const (
...
-			DialContext: (&net.Dialer{
-				Timeout:   timeout,
-				KeepAlive: 2 * nakadiHeartbeatInterval,
-				DualStack: true,
-			}).DialContext,
-			MaxIdleConns:        100,
-			IdleConnTimeout:     2 * nakadiHeartbeatInterval,
-			TLSHandshakeTimeout: timeout,
-		},
+		Transport: &Transport{
```
[the remark this answers is not in the document]
Comment (rdilruba) [reply helper.go]: Yes, that makes sense. I’ve removed tracing for streaming client.

**Comment by stoewer** [comment]
Thank you for this contribution :+1:

**Pull request merged**